tokio-net/src/util/poll_evented.rs

```diff
@@ -342,12 +325,37 @@ where
         Ok(())
     }
 
+    fn new2(io: E, handle: Option<&Handle>) -> io::Result<Self> {
+        let mut pe = Self {
+            io: Some(io),
+            inner: Inner {
+                registration: Registration::default(),
+                read_readiness: AtomicUsize::new(0),
+                write_readiness: AtomicUsize::new(0),
+            },
+        };
+        pe.register(handle)?;
+        Ok(pe)
+    }
+
     /// Ensure that the I/O resource is registered with the reactor.
-    fn register(&self) -> io::Result<()> {
-        self.inner
-            .registration
-            .register(self.io.as_ref().unwrap())?;
-        Ok(())
+    ///
+    /// # Panics
+    ///
+    /// This function panics if `handle` is some but fails to reference a
+    /// reactor.
+    fn register(&mut self, handle: Option<&Handle>) -> io::Result<()> {
+        match handle {
+            Some(handle) => {
+                let handle = handle
+                    .as_priv()
+                    .expect("failed to find reference to reactor");
```

Comment:
Hmm, the function already returns a `Result`, I'm not sure if this should panic or just bail?

Reply:
Good point. Thoughts?

```
let handle = handle.as_priv().ok_or_else(|| {
    io::Error::new(
        io::ErrorKind::NotFound,
        "failed to find reference to reactor",
    )
})?;
```

Reply:
maybe see what @carllerche thinks?

Reply:
If we're assuming that reactor should be always present, then error seems to be unnecessary, and it is exceptional failure.

Reply:
Since a reactor handle cannot be passed in anymore, this was removed in b791b74d14fcad7749106e497edf9e7f6019512d